site/src/pages/TemplateVersionEditorPage/TemplateVersionEditorPage.tsx

```diff
@@ -1,115 +1,359 @@
+
+  useEffect(() => {
+    if (!templateVersionId || !templateVersionStatus) {
+      return;
+    }
+
+    if (templateVersionStatus !== "running") {
+      return;
+    }
+
+    const socket = watchBuildLogsByTemplateVersionId(templateVersionId, {
+      onMessage: (log) => {
+        setLogs((logs) => (logs ? [...logs, log] : [log]));
+      },
+      onDone: async () => {
+        await refetchTemplateVersion();
+      },
+      onError: (error) => {
+        console.error(error);
+      },
+    });
+
+    return () => {
+      socket.close();
+    };
+  }, [refetchTemplateVersion, templateVersionId, templateVersionStatus]);
+
+  return {
+    logs,
+    setLogs,
+  };
+};
+
+const useMissingVariables = (templateVersion: TemplateVersion | undefined) => {
+  const { data: missingVariables } = useQuery({
+    ...templateVersionVariables(templateVersion?.id ?? ""),
+    enabled: templateVersion?.job.error_code === "REQUIRED_TEMPLATE_VARIABLES",
+  });
+  const [isMissingVariablesDialogOpen, setIsMissingVariablesDialogOpen] =
+    useState(false);
+
+  useEffect(() => {
+    if (missingVariables) {
+      setIsMissingVariablesDialogOpen(true);
+    }
+  }, [missingVariables]);
+
+  return {
+    missingVariables,
+    isMissingVariablesDialogOpen,
+    setIsMissingVariablesDialogOpen,
+  };
+};
+
+const generateVersionFiles = async (
+  tarReader: TarReader,
+  fileTree: FileTree,
+) => {
+  const tar = new TarWriter();
+
+  // Add previous non editable files
+  for (const file of tarReader.fileInfo) {
+    if (!isAllowedFile(file.name)) {
+      if (file.type === "5") {
+        tar.addFolder(file.name, {
+          mode: file.mode, // https://github.com/beatgammit/tar-js/blob/master/lib/tar.js#L42
+          mtime: file.mtime,
+          user: file.user,
+          group: file.group,
+        });
+      } else {
+        tar.addFile(file.name, tarReader.getTextFile(file.name) as string, {
+          mode: file.mode, // https://github.com/beatgammit/tar-js/blob/master/lib/tar.js#L42
+          mtime: file.mtime,
+          user: file.user,
+          group: file.group,
+        });
+      }
+    }
+  }
+  // Add the editable files
+  traverse(fileTree, (content, _filename, fullPath) => {
+    // When a file is deleted. Don't add it to the tar.
+    if (content === undefined) {
+      return;
+    }
+
+    if (typeof content === "string") {
+      tar.addFile(fullPath, content);
+      return;
+    }
+
+    tar.addFolder(fullPath);
+  });
+  const blob = (await tar.write()) as Blob;
+  return new File([blob], "template.tar");
+};
+
+const publishVersion = async (options: {
+  version: TemplateVersion;
+  data: PatchTemplateVersionRequest;
+  isActiveVersion: boolean;
+}) => {
+  const { version, data, isActiveVersion } = options;
+  const haveChanges =
+    data.name !== version.name || data.message !== version.message;
+
+  return Promise.all([
+    haveChanges ? patchTemplateVersion(version.id, data) : Promise.resolve(),
```

Comment:
Since it doesn't look like anything is awaiting the results of the `Promise.all` call, I'm wondering if the Promise.all call could save on generating all the extra promises by pushing values to the array directly. That way, the `Promise.all` call doesn't have extra promises to deal with when `haveChanges` or `isActiveVersion` are false